### PR TITLE
Elasticsearch Client Instrumentation

### PR DESCRIPTION
**Pull request by Mpdreamz**
Opening this up to discuss:

Initial implementation of instrumentation of the .NET clients for Elasticsearch:

image

We'll need to discuss what tags and context and types these spans should emit.

### CODE REVIEW

**Comment by Mpdreamz** [comment]
Some action on the oldest PR 😄 

Started on some tests for the elasticsearch integration using `Elasticsearch.Net.VirtualCluster`. 

Currently on prerelease packages, will update this again when we release it to nuget.

**Comment by ejsmith** [comment]
Guess I am pretty surprised that this wasn't one of the first integrations released with the .NET APM agent. I just tried with the `HttpDiagnosticsSubscriber` and it leaves a lot to desire. It just shows an HTTP post with only the host and not the URL (index). Doesn't show the request body even though I know this is a big perf hit and I was under the impression it's turned on for the `HttpDiagnosticsSubscriber` by default. It also shows a really verbose stacktrace with tons of async machinery in it which I thought was gone with .NET Core 3.0 and I know there are libraries or code you can use to clean those stack traces up.

What is the status of adding a 1st class integration with the NEST client?

**Comment by Mpdreamz** [comment]
Thanks for the bump here @ejsmith this has been popping off my queue for too long, while being 90% there for almost a year. Let's get this going. 

Updated the PR with additional tests and marked it as ready for review.

**Comment by Mpdreamz** [comment]
`.ci/linux/build.sh` works on my machine ™️  @gregkalapos is it running in docker on CI, if so NuGet.config from the root needs to be mounted.

**Comment by Mpdreamz** [comment]
ahh i see it is now, wrapped in `dotnet()` in the `JenkinsFile`. Will follow up later.

**Comment by ejsmith** [comment]
@Mpdreamz looks like this is definitely a step in the right direction. Would the metadata for the span include things like the target index name? I realize including the entire query would require a performance overhead, but I'd like to be able to opt into that as well like what is shown here: SQL Statement

**Comment by ejsmith** [comment]
Also, in your example screenshot above it's showing `CallElasticsearch` which is not really useful. I'd like to see it say that it's a POST or a GET with the URL including the index name and a metadata value with the full POST body if I opt into it. I did notice that it appears that the Elasticsearch .NET client does not change the URL path from the query options. I guess it is always setting JSON options in the POST body. I think it would be much better for all kinds of tracing tools if you would put the index into the URL when building the queries.

**Comment by Mpdreamz** [comment]
The updated screenshot is followed: 

> I did notice that it appears that the Elasticsearch .NET client does not change the URL path from the query options. I guess it is always setting JSON options in the POST body. I think it would be much better for all kinds of tracing tools if you would put the index into the URL when building the queries.

The clients actually prefer putting options in the path and querystring unless explicitly opting out, In NEST we instruct the generator to prefer body params for things like fields for various querystring params.

image


Right now we print `.DebugInformation` to the database statement which will preserve the request and response if `.DisableDirectStreaming()` is enabled. This is a bit hacky as the highlighting and formatting is non existent. @gregkalapos thoughts on how to improve here? 

> Would the metadata for the span include things like the target index name? 

Other than being included in the Action name and Statement, we currently do not attach this to the span metadata/labels. This would require some changes in the client to be preserved as separate tokens.


image

**Comment by ejsmith** [comment]
@Mpdreamz that looks awesome! Can't wait to try it out.

File: test/Elastic.Apm.Elasticsearch.Tests/VirtualElasticsearchTests.cs [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+					var searchResponse = await agent.Tracer.CaptureTransaction("Call Client", ApiConstants.ActionExec,
+						async () => await client.SearchAsync<StringResponse>(PostData.Empty)
+					);
+					searchResponse.Should().NotBeNull();
+				}
+				catch (Exception) { }
+
+				var spans = payloadSender.SpansOnFirstTransaction;
+				spans.Should().NotBeEmpty();
+				spans.Should().OnlyContain(s => s.Context.Db.Statement != null);
```
Comment (gregkalapos) [inline test/Elastic.Apm.Elasticsearch.Tests/VirtualElasticsearchTests.cs:77]: This seems failing. 

Debugged this a little bit, here are my findings:

For the last span `RequestPipelineDiagnosticsListener.OnResult` is called with `IApiCallDetails response`  `null`, so we hit [this line](https://github.com/elastic/apm-agent-dotnet/pull/329/files#diff-f77a4d9d52dbf901cdd0736bd4c34521R35) - therefore we don't capture any db info. 

Also I wonder about this case in general, too. Like... we set the `span.name += " (Exception)"` - but is this really an exception? Isn't this more like a case where we just don't get the info we need?

File: src/Elastic.Apm.Elasticsearch/ElasticsearchDiagnosticsListenerBase.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+			span = (Span)ExecutionSegmentCommon.GetCurrentExecutionSegment(_agent).StartSpan(
+				name,
+				ApiConstants.TypeDb,
+				ApiConstants.SubtypeElasticsearch);
+
+			span.Action = name;
+			SetDbContext(span, instance);
+
+			var id = Activity.Current.Id;
```
[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Elastic.Apm.Elasticsearch/ElasticsearchDiagnosticsListenerBase.cs:71]: Aye we call each diagnostic through:

https://github.com/elastic/elasticsearch-net/blob/f1ffbf260132eb83ea4e7950c1bcf19ef679ee6f/src/Elasticsearch.Net/Diagnostics/Diagnostic.cs


e.g: https://github.com/elastic/elasticsearch-net/blob/f1ffbf260132eb83ea4e7950c1bcf19ef679ee6f/src/Elasticsearch.Net/Serialization/DiagnosticsSerializerProxy.cs#L81

Which ensures child activities are started for each new diagnostic event

[the remark this answers is not in the document]
Comment (Mpdreamz) [reply src/Elastic.Apm.Elasticsearch/ElasticsearchDiagnosticsListenerBase.cs:71]: We have docs for this: https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/diagnostic-source.html

Happy to rediscover these myself 😹

[the remark this answers is not in the document]
Comment (gregkalapos) [reply src/Elastic.Apm.Elasticsearch/ElasticsearchDiagnosticsListenerBase.cs:71]: Nice, perfect! 👍

File: test/Elastic.Apm.Elasticsearch.Tests/VirtualElasticsearchTests.cs (hunk 1 above)
Comment (Mpdreamz) [reply test/Elastic.Apm.Elasticsearch.Tests/VirtualElasticsearchTests.cs]: The assertion was flawed we can never capture a statement if response is null, and response is only ever null if an exception occurred in the client. I think the span indicating this is helpful.

**Comment by ejsmith** [comment]
@gregkalapos This is awesome! Any idea on a timeline when this will be in an official nuget release?

**Comment by gregkalapos** [comment]
> @gregkalapos This is awesome! Any idea on a timeline when this will be in an official nuget release?

Can't tell an exact date, but the plan is to include this as part of the next release.

**Comment by ejsmith** [comment]
@gregkalapos is there a CI nuget feed that contains it?

**Comment by ejsmith** [comment]
So @gregkalapos @Mpdreamz I tried pulling the code out for just the Elasticsearch diagnostic listener so I could use it with the currently released client until it is available in an official release. But it turns out that `IDiagnosticListener`, `Span`, `IApmLogger.Scoped`, `ExecutionSegmentCommon` and some others are all internal. Yet every single one of your built in listeners use those. This makes it really hard to create new custom listeners and seems like a really bad thing for users being able to create an ecosystem of listeners and other extensions for this library.

**Comment by ejsmith** [comment]
When you have to cheat and use `InternalsVisibleTo` inside of your own diagnostic subscribers then that makes it really hard for others to contribute and I'm basically at a point where my only option is probably to fork and maintain my own version of the APM client just because I want to enrich my APM data with additional information.

**Comment by Mpdreamz** [comment]
Hey @ejsmith we're moving all of our .NET CI packages to 

https://f.feedz.io/elastic/all/nuget/index.json

Including this repos so the prereleases should be available there 

Re: internal and `InternalsVisibleTo`, I tend to agree and mostly consider them a smell but this was done initially to keep a very strict focus on the public API.

I do think another pass over would be worthwhile, appreciate you re-raising this.

Related:

https://github.com/elastic/apm-agent-dotnet/issues/441
https://github.com/elastic/apm-agent-dotnet/issues/77

But might be worth another ticket to continue to conversation.

**Pull request merged**